### DISCUSSION TITLE
Is VIPS_DISC_THRESHOLD applied for single image? Or entire environment?

### DISCUSSION THREAD

**Discussion by Flash-nDie** [libvips/libvips#3430] [Q&A]
Hi. Sorry for simple question again.
Is VIPS_DISC_THRESHOLD applied for single image? 
I am testing libvips in server application, which is multi-threaded with pthread in C. 
So the application can handle multi image processing with libvips. 
For example, server is resizing three jpg images simultaneously. 
I want to set proper VIPS_DISC_THRESHOLD value. If single images needs 30MB memory and wanting not to use disk, should i set VIPS_DISC_THRESHOLD 30MB(if applied for single image) or 90MB(if applied for entire environment)?

**Comment by jcupitt** [comment]
Hello @Flash-nDie,

Did you see the "how it opens files" chapter? It explains the background:

https://www.libvips.org/API/current/How-it-opens-files.html

tldr: use `thumbnail` and it doesn't matter, it won't use either disc or memory.

**Reply by Flash-nDie** [reply]
@jcupitt 
yes, I read "how it opens files" before and get some information how libvips open files. 
But I wondered the 'VIPS_DISC_THRESHOLD' is set to control whole application memory (declared with VIPS_INIT()), or is set to control for one image. 

By the way, vips_thumbnail() doesn't use memory or disc? Can you explain more?

**Reply by jcupitt** [reply]
`VIPS_DISC_THRESHOLD` sets the memory / disc threshold for the whole of libvips.

It's only used for images which are opened for random access and where the image format does not support random access. `thumbnail` is a sequential operation, so it works by reading the source image a line at a time from top to bottom. It doesn't need to buffer the image in either memory or on disc.

It's a streaming system. You can imagine it like this:

```
file on disc -> 
  decode -> 
    shrink -> 
      encode -> 
        thumbnail on disc
```

All five operations execute at the same time and in parallel, and your pixels flow through the system. No extra memory or disc buffers are needed.

**Reply by Flash-nDie** [reply]
@jcupitt 
oh, thank you for the detailed explanation.
I am using libvips API like vips_image_new_from_buffer (), but thumbnail looks same as you explained. 
I will use thumbnail operation actively. Thanks
